saw-core/src/Verifier/SAW/Simulator/TermModel.hs

```diff
@@ -0,0 +1,1092 @@
+           scRecordType sc fs'
+      VDataType nm ps vs ->
+        do nm' <- traverse loop nm
+           (ps',vs') <- splitAt (length ps) <$> readBackDataTypeParams (primType nm) (ps++vs)
+           scDataTypeAppParams sc nm' ps' vs'
+      VPiType{} ->
+        do (ecs, tm) <- readBackPis tv
+           scGeneralizeExts sc ecs tm
+      VRecursorType d ps m mty ->
+        do d'   <- traverse loop d
+           ps'  <- readBackDataTypeParams (primType d) ps
+           m'   <- readBackValue sc cfg mty m
+           mty' <- loop mty
+           scFlatTermF sc (RecursorType d' ps' m' mty')
+      VTyTerm _s tm ->
+        pure tm
+
+  readBackDataTypeParams (VPiType _nm t body) (v:vs) =
+    do v' <- readBackValue sc cfg t v
+       t' <- applyPiBody body (ready v)
+       vs' <- readBackDataTypeParams t' vs
+       return (v':vs')
+
+  readBackDataTypeParams _ [] = return []
+  readBackDataTypeParams _ _ = panic "readBackTValue" ["Arity mismatch in data type in readback"]
+
+  readBackPis (VPiType nm t pibody) =
+    do t' <- loop t
+       ec <- scFreshEC sc nm t'
+       ?recordEC ec
+       ecTm <- scExtCns sc ec
+       ecVal <- delay (reflectTerm sc cfg t ecTm)
+       body <- applyPiBody pibody ecVal
+       (ecs,body') <- readBackPis body
+       return (ec:ecs, body')
+
+  readBackPis t =
+    do tm <- loop t
+       return ([], tm)
+
+evalType :: Sim.SimulatorConfig TermModel -> Term -> IO (TValue TermModel)
+evalType cfg tm =
+  Sim.evalSharedTerm cfg tm >>= \case
+    TValue tv -> pure tv
+    _ -> panic "evalType" ["Expected type value"]
+
+reflectTerm ::
+  (?recordEC :: BoundECRecorder) =>
+  SharedContext ->
+  Sim.SimulatorConfig TermModel ->
+  TValue TermModel ->
+  Term {- ^ closed term to reflect -} ->
+  IO (Value TermModel)
+reflectTerm sc cfg = loop
+  where
+  loop tv tm = case tv of
+    VUnitType -> pure VUnit
+    VBoolType -> return (VBool (Left tm))
+    VIntType  -> return (VInt (Left tm))
+    VIntModType m -> return (VIntMod m (Left tm))
+    VVecType n VBoolType -> return (VWord (Left (n,tm)))
+    VArrayType a b ->
+      do a' <- readBackTValue sc cfg a
+         b' <- readBackTValue sc cfg b
+         return (VArray (TMArray a a' b b' tm))
+
+    VSort s -> pure (TValue (VTyTerm s tm))
+
+    VPiType nm t pibody ->
+      return (VFun nm (\x ->
+        do v <- force x
+           tbody <- applyPiBody pibody x
+           xtm <- readBackValue sc cfg t v
+           body <- scApply sc tm xtm
+           reflectTerm sc cfg tbody body
+        ))
+
+    -- TODO: eta-expanding vectors like this kinda sucks.  It would be
+    -- better if we could treat them abstractly, as we are doing below
+    -- with records, pairs and data types.  However, that would be
+    -- a pretty invasive change in the primitives.
+    VVecType n t ->  -- NB: t /= Bool
+      do a  <- readBackTValue sc cfg t
+         n' <- scNat sc n
+         vs <- V.generateM (fromIntegral n) $ \i -> -- TODO fromIntegral...
```

Comment:
I don't know if it would be easy to avoid these uses of `fromIntegral`, but I thought I'd point them out since you have a TODO comment there.

Reply:
The main problem is that the `Vector` module expects `Int` for everything, so we have to do the projection somewhere.  Ultimately, I think we'll solve this by using the `SeqMap` type from `Cryptol` instead, but that is a little down the road.